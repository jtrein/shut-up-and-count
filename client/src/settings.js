export const ALLOWED_MIME = 'text/plain';
export const API_HOST = process.env.API_HOST || 'http://localhost';
export const API_PORT = process.env.API_PORT || 4000;
export const ERR_FILE_SIZE = 'File is too big.';
export const ERR_FILE_TYPE = 'File is not a .txt file.';
export const MAX_FILE_SIZE = 10 * 1024 * 1024;
export const MSG_SERVER_ERR = 'Sorry, it appears our server is gone.';
export const PARSE_ENDPOINT = `${API_HOST}:${API_PORT}/parse`;
export const UPLOAD_KEY = 'textFile';
