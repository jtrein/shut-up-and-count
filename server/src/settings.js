module.exports = {
  IS_DEV: process.env.NODE_ENV === 'development',
  fileKey: 'textFile',
};
