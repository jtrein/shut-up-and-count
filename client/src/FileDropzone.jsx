import React, { Component } from 'react';
import Dropzone from 'react-dropzone';

import { ALLOWED_MIME, MAX_FILE_SIZE } from './settings';
import dropzoneStyles from './FileDropzone.styles';

class FileDropzone extends Component {
  constructor() {
    super();

    this.state = {
      files: [],
      dropzoneActive: false,
    };
  }

  onDragEnter() {
    this.setState({
      dropzoneActive: true,
    });
  }

  onDragLeave() {
    this.setState({
      dropzoneActive: false,
    });
  }

  onDrop(acceptedFiles) {
    this.setState({
      files: acceptedFiles,
      dropzoneActive: false,
    });
  }

  render() {
    const { files, dropzoneActive } = this.state;

    return (
      <Dropzone
        accept={ALLOWED_MIME}
        multiple={false}
        maxSize={MAX_FILE_SIZE}
        onDrop={(acceptedFiles, rejectedFiles) => {
          this.onDrop(acceptedFiles, rejectedFiles);
        }}
        onDragEnter={() => this.onDragEnter()}
        onDragLeave={() => this.onDragLeave()}
        style={dropzoneStyles.dropzone}
      >
        {/* overlay */}
        { dropzoneActive && <div style={dropzoneStyles.overlay}>Go ahead, drop it!</div> }

        {/* list files */}
        {
          files.map(({ name, size }) => (
            <p key={`${name}-${size}`} style={dropzoneStyles.textHighlight}>
              {name}
            </p>
          ))
        }

        <p>Click or drop .txt files.</p>
        <small>(max size is 10MB)</small>
      </Dropzone>
    );
  }
}

export default FileDropzone;
