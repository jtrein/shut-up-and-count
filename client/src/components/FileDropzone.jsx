import React, { Component } from 'react';
import { func } from 'prop-types';
import Dropzone from 'react-dropzone';

import { ALLOWED_MIME, MAX_FILE_SIZE } from '../settings';
import dropzoneStyles from '../styles/FileDropzone.styles';

const getApproxFileSize = kb => (kb < 1 ? '~1' : kb);

class FileDropzone extends Component {
  constructor() {
    super();

    this.state = {
      file: [],
      dropzoneActive: false,
    };
  }

  onDragEnter() {
    this.setState({
      dropzoneActive: true,
    });
  }

  onDragLeave() {
    this.setState({
      dropzoneActive: false,
    });
  }

  onDrop(acceptedFile) {
    this.setState({
      file: acceptedFile,
      dropzoneActive: false,
    }, () => {
      if (acceptedFile.length) this.props.onDropAccepted(acceptedFile);
    });
  }

  onDropRejected(rejectedFile) {
    const { size, type } = rejectedFile[0];
    if (size > MAX_FILE_SIZE) {
      this.props.onDropRejected('size');
    }
    if (type !== ALLOWED_MIME) {
      this.props.onDropRejected('type');
    }
  }

  render() {
    const { file, dropzoneActive } = this.state;
    const { name, size } = file.length ? file[0] : [];

    return (
      <Dropzone
        accept={ALLOWED_MIME}
        multiple={false}
        maxSize={MAX_FILE_SIZE}
        onDrop={acceptedFile => this.onDrop(acceptedFile)}
        onDropRejected={rejectedFile => this.onDropRejected(rejectedFile)}
        onDragEnter={() => this.onDragEnter()}
        onDragLeave={() => this.onDragLeave()}
        style={dropzoneStyles.dropzone}
      >
        {/* overlay */}
        { dropzoneActive && <div style={dropzoneStyles.overlay}>Go ahead, drop it!</div> }

        {/* file info */}
        {file.length ?
          (
            <p style={dropzoneStyles.textHighlight}>
              {name} &ndash; {getApproxFileSize(Math.floor(size / 1024))}kb
            </p>
          ) : null
        }

        <p>Click or drop .txt files.</p>
        <small>(max size is 10MB)</small>
      </Dropzone>
    );
  }
}

FileDropzone.propTypes = {
  onDropAccepted: func.isRequired,
  onDropRejected: func.isRequired,
};

export default FileDropzone;
