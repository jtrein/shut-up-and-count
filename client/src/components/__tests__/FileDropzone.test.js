/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Dropzone from 'react-dropzone';
import toJSON from 'enzyme-to-json';

import FileDropzone from '../FileDropzone';
import { ALLOWED_MIME, MAX_FILE_SIZE } from '../../settings';

Enzyme.configure({ adapter: new Adapter() });

test('renders without crashing', () => {
  const div = document.createElement('div');
  render(<FileDropzone onDropAccepted={() => {}} />, div);
  unmountComponentAtNode(div);
});

test('renders expected initial components', () => {
  const wrapper = mount(<FileDropzone onDropAccepted={() => {}} />);
  expect(wrapper.find('p').text()).toBe('Click or drop .txt files.');
  expect(wrapper.find(Dropzone).length).toBe(1);
});

test('renders expected text', () => {
  const wrapper = mount(<FileDropzone onDropAccepted={() => {}} />);
  expect(wrapper.find('p').text()).toBe('Click or drop .txt files.');
  expect(wrapper.find('small').text()).toBe('(max size is 10MB)');
});

test('renders expected props from settings', () => {
  const wrapper = mount(<FileDropzone onDropAccepted={() => {}} />);
  expect(wrapper.find(Dropzone).props().accept).toBe(ALLOWED_MIME);
  expect(wrapper.find(Dropzone).props().maxSize).toBe(MAX_FILE_SIZE);
});

test('renders expected overlay text when active', () => {
  const wrapper = mount(<FileDropzone onDropAccepted={() => {}} />);
  wrapper.setState({ dropzoneActive: true });
  expect(wrapper.find('div').first().text()).toContain('Go ahead, drop it!');
});

test('renders file info text when file', () => {
  const fileFixture = [{ name: 'hello', size: 123456 }];
  const wrapper = mount(<FileDropzone onDropAccepted={() => {}} />);
  wrapper.setState({ file: fileFixture });
  expect(wrapper.find('div').first().text()).toContain('hello');
  expect(wrapper.find('div').first().text()).toContain('120kb');
});

test('dropzoneActive state is true on dragenter', () => {
  const wrapper = mount(<FileDropzone onDropAccepted={() => {}} />);
  wrapper.simulate('dragenter');
  expect(wrapper.state().dropzoneActive).toBe(true);
});

test('dropzoneActive state is false on dragleave', () => {
  const wrapper = mount(<FileDropzone onDropAccepted={() => {}} />);
  wrapper.simulate('dragleave');
  expect(wrapper.state().dropzoneActive).toBe(false);
});

test('proper state changes on drop', () => {
  const spy = jest.fn();
  const wrapper = mount(<FileDropzone onDropAccepted={spy} />);
  wrapper.simulate('drop');
  expect(wrapper.state().dropzoneActive).toBe(false);
  expect(spy.mock.calls.length).toBe(1);
});

test('Snapshot: renders as expected', () => {
  const wrapper = mount(<FileDropzone onDropAccepted={() => {}} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
