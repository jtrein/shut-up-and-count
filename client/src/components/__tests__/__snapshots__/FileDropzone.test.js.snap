// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: renders as expected 1`] = `
<FileDropzone
  onDropAccepted={[Function]}
  onDropRejected={[Function]}
>
  <t
    accept="text/plain"
    disableClick={false}
    disablePreview={false}
    disabled={false}
    maxSize={10485760}
    minSize={0}
    multiple={false}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDrop={[Function]}
    onDropRejected={[Function]}
    preventDropOnDocument={true}
    style={
      Object {
        "border": "10px dashed #EEE",
        "fontSize": 24,
        "fontWeight": 700,
        "margin": 24,
        "maxWidth": 500,
        "padding": 72,
        "textAlign": "center",
      }
    }
  >
    <div
      aria-disabled={false}
      className=""
      onClick={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "border": "10px dashed #EEE",
          "fontSize": 24,
          "fontWeight": 700,
          "margin": 24,
          "maxWidth": 500,
          "padding": 72,
          "textAlign": "center",
        }
      }
    >
      <p>
        Click or drop .txt files.
      </p>
      <small>
        (max size is 10MB)
      </small>
      <input
        accept="text/plain"
        autoComplete="off"
        disabled={false}
        multiple={false}
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
    </div>
  </t>
</FileDropzone>
`;
